### PR TITLE
[ir] Small optimizations to codegen

### PR DESCRIPTION
**Pull request by bobcao3**
1. Implement proper bulk `insert` in Block, which utilizes the STL optimized bulk insert operation instead of repeatedly modifying the same vector. (therefore also improves replacing single statements with a `VecStatement`)
2. (Should consider whether we should use this or not) Use `llvm::SmallVector`. A bit of refactoring is done so that this can be easily toggled between STL and LLVM ADT.

Under DEBUG build options the `compile_to_offload` time of `cornell_box` reduces from 18.5s to 17.4s. Most of this is though the `SmallVector`.

Under release flag performance difference seems hard to measure.

During profiling, it seems that in these complex kernels, I saw `28.26%` used by `whole_kernel_cse`, and `66.88%` used by `cfg_optimization`. Across the optimization, one hot spot is `replace_all_usages_with` which is called thousands of times resulting in quite a big total amount of time spent on it. However the time spent on `replace_all_usages_with` is mostly IR traversal (i.e. `accept()` ). Only good way to optimize this is to somehow restructure IR to achieve smaller memory footprint and less indirections.

### CODE REVIEW

**Comment by bobcao3** [comment]
Possible pain point: LLVM small vector is a heavy dependency. Need to consider this carefully. (Switching it to std::vector is simple enough tho, there's a `using` statement in there.

**Comment by bobcao3** [comment]
> > Only good way to optimize this is to somehow restructure IR to achieve smaller memory footprint and less indirections.
> 
> I wonder if implementing our own slab allocator/arena would help this problem.

Can we quickly test that with a third party library?

**Comment by k-ye** [comment]
> Can we quickly test that with a third party library?

https://github.com/google/benchmark

**Comment by bobcao3** [comment]
> > Can we quickly test that with a third party library?
> 
> https://github.com/google/benchmark

I mean allocators library out there

File: taichi/ir/ir.h [hunk 1]
```diff
@@ -98,9 +101,12 @@ class Identifier {
   }
 };
 
+using stmt_vector = llvm::SmallVector<pStmt, 8>;
```
[the remark this answers is not in the document]
Comment (k-ye) [reply taichi/ir/ir.h:104]: Then let's just remove it later

**Comment by k-ye** [comment]
> I mean allocators library out there

Not sure, either. I was just thinking whether a dead simple linear allocator could provide better locality ¯\_(ツ)_/¯

**Pull request merged**